Add ValidationRequiredError error context

aeneasr opened the pull request.

santhosh-tekuri commented on errors_validation.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+// ValidationRequiredError is used when a
+type ValidationRequiredError struct {
```
name ends with 'Error' but does not implement error interface

aeneasr replied:
True, `ValidationErrorRequired` or `ValidationContextRequired` is probably a better naming!

santhosh-tekuri commented on errors_validation.go:
```diff
@@ -0,0 +1,47 @@
+	"strings"
+)
+
+type validationErrorContext interface {
```
how user define extensions provide error context ?

aeneasr replied:
I didn't have a chance yet to check out the user defined extensions introduced by you recently.

santhosh-tekuri commented on errors.go:
```diff
@@ -50,6 +50,9 @@ type ValidationError struct {
 	SchemaPtr string
 
+	// Context represents error context for this specific validation error.
+	Context validationErrorContext
```
field is exported, but not its type

santhosh-tekuri replied:
SchemaPtr and InstancePtr are already providing error context. Adding another field named Context is misleading.
may be `Details` is better name

aeneasr replied:
The only other logical here is `interface{}`, but using the private type helps avoid some internal type assertions. Usage of this field would be, for example:

```go
			if err.Context == nil {
				// no context
			}
			switch ctx := err.Context.(type) {
			case *jsonschema.ValidationRequiredError:
				for _, required := range ctx.MissingPropertyPtrs {
					// do whatever
				}
			}
```

aeneasr replied:
I agree, `Context` is not very useful. Details would be indeed better.

santhosh-tekuri commented on errors_validation.go:
```diff
@@ -0,0 +1,47 @@
+
+// ValidationRequiredError is used when a
+type ValidationRequiredError struct {
+	MissingPropertyPtrs []string
```
lengthy name. required keyword works only for properties. so including `Property` in name feels like stutter. May be `MissingPtrs` or just `Missing` would be concise.

aeneasr commented:
@santhosh-tekuri so what do you think of the general approach taken here? Is adding a Detail field something you would consider upstream? I think it adds a lot of usability to the library. Some more error types we'd like to introduce are:

- `minLength`/`maxLength`: a struct with `{ MinLength int, ActualLength int }`
- `format`: a struct with `{ Pattern string }`
- ...

santhosh-tekuri commented on schema.go:
```diff
@@ -278,11 +278,11 @@ func (s *Schema) validate(v interface{}) error {
 			var missing []string
 			for _, pname := range s.Required {
 				if _, ok := v[pname]; !ok {
-					missing = append(missing, strconv.Quote(pname))
```
pname may contain comma or space. because of this it was quoted in message

aeneasr replied:
We want the raw JSON pointers here (e.g. `#/foo/bar`) because we want to pass that to the error details. The error message itself is still being quoted: https://github.com/ory/jsonschema/pull/1/files#diff-02934e0fdc85f35aa62e19cf7c6f6bcaR39

santhosh-tekuri replied:
got it. but before including pname in json pointer, some special characters needs to be escaped.
check https://github.com/santhosh-tekuri/jsonschema/blob/master/schema.go#L557

aeneasr replied:
Right! I though [`joinPtr`](https://github.com/ory/jsonschema/pull/1/files#diff-02934e0fdc85f35aa62e19cf7c6f6bcaR21) would do that but I assumed that incorrectly.

aeneasr commented:
@santhosh-tekuri I've addressed your comments and will go ahead with running this as a fork. Let me know if you'd want these changes upstream (I think people would definitely appreciate that) and I'll submit a PR.
